### PR TITLE
Base images with CUDA support, build tooling improvements, and explicit dev tools

### PR DESCRIPTION
**Pull request by randalburns**
This PR adds new bases for CUDA support. It also updates all existing bases to explicitly call out all dev tools. New build automation was added to make it easier to build multiple images, along with a simple templating framework for re-use of Dockerfiles when building bases with minor differences.

### CODE REVIEW

File: base.py [hunk 1]
```diff
@@ -39,54 +50,122 @@ def _generate_image_tag_suffix(self) -> str:
...
-        [print(ln[list(ln.keys())[0]], end='') for ln in client.api.build(path=build_dir,
-                                                                          tag=named_tag,
-                                                                          nocache=no_cache,
-                                                                          pull=True, rm=True,
-                                                                          decode=True)]
+        build_args = None
+        if os.path.isfile(os.path.join(base_dir, "dockerfile_template.json")):
+            with open(os.path.join(base_dir, "dockerfile_template.json"), 'rt') as tf:
+                template_data = json.load(tf)
+            build_dir = os.path.join(self._get_root_dir(), '_templates', template_data['template'])
```
Comment (davclark) [inline base.py]: This works for now, but what about build contexts with extra files (e.g., the way I currently manage the R build process)? Perhaps we should just construct a temporary build dir?

Comment (dkleissa) [reply base.py]: For that case I would say extra files should go with the Dockerfile and additional `ARG` instructions added to switch between files as needed.

File: _templates/python3-minimal-template/Dockerfile [hunk 2]
```diff
@@ -32,45 +36,42 @@ ENV CONDA_DIR=/opt/conda \
     GIGANTUM_SHARE=/mnt/share \
     GIGANTUM_WORKDIR=/mnt/labbook \
     SHELL=/bin/bash \
-    MINICONDA_VERSION=4.5.4 \
-    MINICONDA_MD5=a946ea1d0c4a642ddf0c3a26a18bb16d \
+    MINICONDA_VERSION=4.5.12 \
```
[the remark this answers is not in the document]
Comment (dkleissa) [reply _templates/python3-minimal-template/Dockerfile]: Oh good catch. I think probably fine? Should be a graceful upgrade? I think worth it for the alleged conda speed ups.

File: python3-minimal-cuda-10.0/dockerfile_template.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+  "template": "python3-minimal-template",
+  "args": {
+    "BASE_IMAGE": "nvidia/cuda:10.0-cudnn7-runtime-ubuntu18.04"
```
Comment (davclark) [inline python3-minimal-cuda-10.0/dockerfile_template.json]: Here, and elsewhere, the use of the runtime images creates a high barrier to doing serious development on the GPU. Anyone with a reasonable GPU for CUDA is going to have a "nice" computer. This entails they probably have enough space to eat the additional dependencies from cuda development libraries.

As it stands, I am unable to get libcudnn7-dev installed on the CUDA 8 images easily - the system wants to install libraries that are up-to-date for CUDA-10.

Comment (davclark) [reply python3-minimal-cuda-10.0/dockerfile_template.json]: OK - willing to leave this in a "wait for users to tell us its broken" state.

**Pull request merged**